app/lib/shared/rediscache.dart

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:simple_cache/simple_cache.dart';
+import 'package:gcloud/service_scope.dart' as ss;
+import 'package:gcloud/db.dart' as db;
+import 'package:appengine/appengine.dart';
+import 'package:logging/logging.dart';
+
+import '../frontend/models.dart' show Secret, SecretKey;
+import 'versions.dart';
+
+final Logger _log = new Logger('rediscache');
+
+/// The active cache.
+///
+/// Can only be used within the context of [withAppEngineAndCache].
+Cache<List<int>> get cache => ss.lookup(#_cache) as Cache<List<int>>;
+
+/// Run [fn] with AppEngine services and [cache].
+///
+/// See `package:appengine` for details on how [withAppEngineServices] sets up
+/// services like datastore and logging.
+Future withAppEngineAndCache(FutureOr Function() fn) async {
+  return withAppEngineServices(() async {
+    // Use in-memory cache, if not running on AppEngine
+    if (Platform.environment.containsKey('GAE_VERSION')) {
+      return await _withRedisCache(fn);
+    } else {
+      _log.warning('using in-memory cache instead of redis');
+      return await _withInmemoryCache(fn);
+    }
+  });
+}
+
+/// Run [fn] with [cache] connected to a redis cache.
+Future _withRedisCache(FutureOr Function() fn) async {
+  // TODO: use in-memory if GAE_VERSION is not set...
+  final s = await db.dbService.lookup([
+    db.dbService.emptyKey.append(Secret, id: SecretKey.redisConnectionString),
+  ]);
+  // Validate that we got a connection string
+  if (s.isEmpty || !(s[0] is Secret) || (s[0] as Secret).value.isEmpty) {
+    throw Exception('Secret ${SecretKey.redisConnectionString} is missing');
+  }
+  final connectionString = (s[0] as Secret).value;
+
+  // Create and register a cache
+  final cacheProvider = Cache.redisCacheProvider(connectionString);
+  ss.register(#_cache, Cache(cacheProvider));
```

Comment:
I hate it for the boilerplate, but I'd prefer to use the `registerCache(Cache cache)` pattern here too, as it is easier to scan for its uses.

Reply:
I could care less, it's just internal boiler plate...

Reply:
besides the symbol is internal, so it can be set outside this file...